bin/generated_template.d.ts

```diff
@@ -0,0 +1,113 @@
+/** Provides information pointing to a location within a source. */
+export interface Location {
+  /** Line in the parsed source (1-based). */
+  line: number;
+  /** Column in the parsed source (1-based). */
+  column: number;
+  /** Offset in the parsed source (0-based). */
+  offset: number;
+}
+
+/** The `start` and `end` position's of an object within the source. */
+export interface LocationRange {
+  /** Any object that was supplied to the `parse()` call as the `grammarSource` option. */
+  source: any;
```

Comment:
Don't worry about being too pedantic.  I indeed did ask for it. :)

Yeah, the comment is wrong.  `grammarSource` is often a string containing the file name, but could also be an object that produces a useful string when `String(grammarSource)` is applied, e.g. something with a `toString()` method.  They also need to be testable for equality with `===`.  As such, numbers would work also, as long as they aren't NaN.

Reply:
The GrammarLocation class in lib/grammar-location.js is a good example of why you might want an object as the grammarSource.

Reply:
And you're right that `GrammarLocation` has a few methods that the generated parser can use if they are there.  They should be documented.

Reply:
Yeah so my final type; `type Source = File | string | {} | null | undefined;` in the last comment may be close to what you want. It'll allow literally anything with some intellisense hints about common uses.

But here's another try at it based upon your new constraints:
```ts
// I'm sure you can come up with a better name for this
// This will contain all of the methods you mentioned that the parser can use if they're there.
interface ObjectSource {
    ... // Methods go here, you can make them required if you want as `AnyObject` will catch it in.
}

// `AnyObject` is a bit broader than you talked about; there's no guarantee it has a useful `toString` method. You might think that you could fix this with the type `{ toString(): string }` but unfortunately `Object` itself defines a `toString` method that returns `"[object Object]"`
// `
//
// You can probably decide for yourself if `null` and `undefined` makes sense. They most likely make sense to mean "not present" rather than a valid value though.
// It sounds like arrays probably don't make sense here. If they do you could add `readonly unknown[]` to the union to allow any array (readonly or not)
// `unique symbol` - the return of `Symbol(...)`  may not be usable because of the `===`/`toString` constraints?
// `symbol` - the return of `Symbol.for()` may make more sense but still a bit strange to put here.
// It sounds like functions almost certainly don't make sense here.
type GrammarSource = ObjectSource | AnyObject | string | number | bigint;
```

Notably this new version does _NOT_ just allow anything but I left a comment about everything it omits.

Reply:
I've got this:

```ts
export interface GrammarSourceObject {
  readonly source?: undefined | string | unknown;
  readonly offset?: undefined | ((loc: Location) => void);
  readonly toString: () => string;
}

export type GrammarSource = string | GrammarSourceObject | unknown;
```

That `unknown` is too broad, and I'd be open to just removing it in favor of the things we know work and are useful.

Reply:
Alright, what things are known to work and are useful? I can help you figure that out.

Reply:
I'm driving across Kansas today, currently charging the car in Maple Hill.  Sorry in advance for spotty connectivity.

I think I'm probably the only one who ever uses anything other than a string, and the only other thing I ever use is GrammarLocation, which conforms to GrammarSourceObject.  Let's just take the unknown out and let people file bugs when they want more. :)